### PR TITLE
Add additional query time statistics

### PR DESCRIPTION
**Pull request by tanjialiang**
Resource waiting time currently represents 3 big chucks:
1. waiting for cluster to meet minimum number of healthy nodes requirement.
2. transforming AST to Analysis representation (Plus some optimization)
3. ACL checking

It makes more sense to break the stage up to 3 separate stages to reveal more details.

Test plan - Changed current unit test to include the testing of the changes.

```
== NO RELEASE NOTE ==
```

depends on https://github.com/facebookexternal/presto-facebook/pull/1570

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/dispatcher/LocalDispatchQuery.java [hunk 1]
```diff
@@ -173,7 +173,11 @@ private void waitForMinimumWorkers()
     {
         ListenableFuture<?> minimumWorkerFuture = clusterSizeMonitor.waitForMinimumWorkers();
         // when worker requirement is met, wait for query execution to finish construction and then start the execution
-        addSuccessCallback(minimumWorkerFuture, () -> addSuccessCallback(queryExecutionFuture, this::startExecution));
+        addSuccessCallback(minimumWorkerFuture, () -> {
+            if (stateMachine.transitionToSemanticAnalyzing()) {
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/dispatcher/LocalDispatchQuery.java:177]: Is this `if statement` needed? Will we go into a limbo state (no execution getting started) if this returns false?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java [hunk 2]
```diff
@@ -180,10 +180,12 @@ private SqlQueryExecution(
                     preparedQuery.getParameters(),
                     warningCollector);
 
-            this.analysis = analyzer.analyze(preparedQuery.getStatement());
-
+            this.analysis = analyzer.semanticAnalyze(preparedQuery.getStatement(), false);
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java:183]: nit: `analyzeSemantic`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 3]
```diff
@@ -75,14 +75,28 @@ public Analysis analyze(Statement statement)
     }
 
     public Analysis analyze(Statement statement, boolean isDescribe)
+    {
+        Analysis analysis = semanticAnalyze(statement, isDescribe);
+        aclCheck(analysis);
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:80]: nit: `checkAccessControl`?

**Review by shixuan-fan — COMMENTED** [review]
Some nits and one question

**Comment by tanjialiang** [comment]
Actually this PR created a big race condition.. 
waitingForResource and semantic analysis is executed parallely.. We cannot transit state like this..

**Comment by tanjialiang** [comment]
> Actually this PR created a big race condition..
> waitingForResource and semantic analysis is executed parallely.. We cannot transit state like this..

Actually semantic analysis and acl checking start very early, even before queuing/waitingForPrerequisite

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 4]
```diff
@@ -75,14 +75,28 @@ public Analysis analyze(Statement statement)
...
         analyzeForUtilizedColumns(analysis, analysis.getStatement());
+        return analysis;
+    }
 
-        // check column access permissions for each table
+    /**
+     * check column access permissions for each table
+     * @param analysis the Analysis that needs to check ACL for
+     */
+    public void aclCheck(Analysis analysis)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:98]: Call it `checkAccessControl`. We don't usually use acronym or abbreviation in our codebase.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 5]
```diff
@@ -75,14 +75,28 @@ public Analysis analyze(Statement statement)
...
         Analysis analysis = new Analysis(rewrittenStatement, parameters, isDescribe);
         StatementAnalyzer analyzer = new StatementAnalyzer(analysis, metadata, sqlParser, accessControl, session, warningCollector);
         analyzer.analyze(rewrittenStatement, Optional.empty());
         analyzeForUtilizedColumns(analysis, analysis.getStatement());
+        return analysis;
+    }
 
-        // check column access permissions for each table
+    /**
+     * check column access permissions for each table
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:95]: `AccessControl` has a lot of method to check ACL at different places. Not quite sure if we call it ACL checking here is precise. Will leave it to @shixuan-fan for further design decision making :D

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateTimer.java [hunk 6]
```diff
@@ -250,6 +289,16 @@ public Duration getResourceWaitingTime()
         return getDuration(resourceWaitingTime, beginResourceWaitingNanos);
     }
 
+    public Duration getSemanticAnalyzingTime()
+    {
+        return getDuration(semanticAnalyzingTime, beginSemanticAnalyzingNanos);
+    }
+
+    public Duration getAclCheckingTime()
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateTimer.java:297]: `getAccessControlCheckingTime`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java (hunk 5 above)
Comment (shixuan-fan) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:95]: This is a good point, we should probably call this `checkColumnAccessPermissions` :)

Comment (tanjialiang) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:95]: > This is a good point, we should probably call this `checkColumnAccessPermissions` :)

Thanks! Good to know. I think I'll rename the Scuba column as well to this name .

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 7]
```diff
@@ -75,14 +75,28 @@ public Analysis analyze(Statement statement)
     }
 
     public Analysis analyze(Statement statement, boolean isDescribe)
+    {
+        Analysis analysis = analyzeSemnatic(statement, isDescribe);
+        checkColumnAccessPermissions(analysis);
+        return analysis;
+    }
+
+    public Analysis analyzeSemnatic(Statement statement, boolean isDescribe)
```
Comment (shixuan-fan) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java:84]: nit: typo `analyzeSemantic`

**Review by shixuan-fan — APPROVED** [review]
One typo. Otherwise LGTM

**Comment by shixuan-fan** [comment]
This is one test failure which might be related:

```
presto-master_1       | 2021-06-16T07:09:39.867+0545	ERROR	dispatcher-query-104	com.google.common.util.concurrent.AbstractFuture	RuntimeException while executing runnable CallbackListener{com.facebook.airlift.concurrent.MoreFutures$4@581b71bb} with executor MoreExecutors.directExecutor()
presto-master_1       | java.lang.IllegalStateException: Start time not set
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.nanosSince(QueryStateTimer.java:350)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.beginDispatching(QueryStateTimer.java:169)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.beginPlanning(QueryStateTimer.java:180)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.beginStarting(QueryStateTimer.java:192)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.beginRunning(QueryStateTimer.java:203)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.beginFinishing(QueryStateTimer.java:213)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.endQuery(QueryStateTimer.java:224)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateTimer.endQuery(QueryStateTimer.java:219)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateMachine.transitionToFailed(QueryStateMachine.java:747)
presto-master_1       | 	at com.facebook.presto.execution.QueryStateMachine.transitionToFailed(QueryStateMachine.java:735)
presto-master_1       | 	at com.facebook.presto.dispatcher.LocalDispatchQuery.lambda$new$0(LocalDispatchQuery.java:97)
presto-master_1       | 	at com.facebook.airlift.concurrent.MoreFutures$4.onFailure(MoreFutures.java:572)
presto-master_1       | 	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1052)
presto-master_1       | 	at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:398)
presto-master_1       | 	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1029)
presto-master_1       | 	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:871)
presto-master_1       | 	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:716)
presto-master_1       | 	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.afterRanInterruptibly(TrustedListenableFutureTask.java:133)
presto-master_1       | 	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:80)
presto-master_1       | 	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
presto-master_1       | 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
presto-master_1       | 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
presto-master_1       | 	at java.lang.Thread.run(Thread.java:748)
presto-master_1       | 
presto-master_1       | 
java.lang.RuntimeException: error invoking methods annotated with io.prestodb.tempto.AfterTestWithContext
	at io.prestodb.tempto.internal.initialization.TestInitializationListener.invokeMethodsAnnotatedWith(TestInitializationListener.java:308)
	at io.prestodb.tempto.internal.initialization.TestInitializationListener.runAfterWithContextMethods(TestInitializationListener.java:297)
	at io.prestodb.tempto.internal.initialization.TestInitializationListener.onTestFinished(TestInitializationListener.java:273)
	at io.prestodb.tempto.internal.initialization.TestInitializationListener.onTestFailure(TestInitializationListener.java:256)
	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1716)
	at org.testng.internal.Invoker.runTestListeners(Invoker.java:1699)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:703)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.prestodb.tempto.internal.initialization.TestInitializationListener.invokeMethodsAnnotatedWith(TestInitializationListener.java:305)
	... 13 more
Caused by: java.lang.RuntimeException: Child process didn't finish within given timeout
	at io.prestodb.tempto.process.LocalCliProcess.waitForWithTimeoutAndKill(LocalCliProcess.java:42)
	at io.prestodb.tempto.internal.process.CliProcessBase.waitForWithTimeoutAndKill(CliProcessBase.java:148)
	at com.facebook.presto.tests.cli.PrestoCliLauncher.stopPresto(PrestoCliLauncher.java:63)
	at com.facebook.presto.tests.cli.PrestoCliTests.stopPresto(PrestoCliTests.java:92)
	... 18 more
2021-06-16 07:14:58 INFO: FAILURE     /    com.facebook.presto.tests.cli.PrestoCliTests.shouldHandleTransaction (Groups: cli) took 5 minutes and 20 seconds
2021-06-16 07:14:58 SEVERE: Failure cause:
org.testng.internal.thread.ThreadTimeoutException: Method com.facebook.presto.tests.cli.PrestoCliTests.shouldHandleTransaction() didn't finish within the time-out 300000
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:255)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:284)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.Reader.read(Reader.java:100)
	at java.util.Scanner.readInput(Scanner.java:804)
	at java.util.Scanner.hasNext(Scanner.java:1446)
	at io.prestodb.tempto.internal.process.CliProcessBase.hasNextOutput(CliProcessBase.java:85)
	at com.facebook.presto.tests.cli.PrestoCliProcess.readLinesUntilPrompt(PrestoCliProcess.java:37)
	at com.facebook.presto.tests.cli.PrestoCliTests.shouldHandleTransaction(PrestoCliTests.java:245)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by shixuan-fan** [comment]
Some related failures:

test-other-modules:
`com.fasterxml.jackson.databind.exc.ValueInstantiationException: Cannot construct instance of `com.facebook.presto.execution.QueryStats`, problem: semanticAnalyzingTime is null`

presto-main:
```
Error:  Failures: 
Error:    TestQueryStats.testJson:243->assertExpectedQueryStats:256 expected [32.00ns] but found [15.00ns]
```

**Pull request merged**